### PR TITLE
Refactor DOH classes to Kotlin

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt [hunk 1]
```diff
@@ -13,33 +13,27 @@
...
 
 /**
  * Internal Bootstrap DNS implementation for handling initial connection to DNS over HTTPS server.
  *
  * Returns hardcoded results for the known host.
  */
-final class BootstrapDns implements Dns {
-  private final String dnsHostname;
-  private final List<InetAddress> dnsServers;
+internal class BootstrapDns(private val dnsHostname: String, private val dnsServers: List<InetAddress>) : Dns {
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt]: nit: wrap these?

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt [hunk 2]
```diff
@@ -13,33 +13,27 @@
...
-
-  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
-    if (!this.dnsHostname.equals(hostname)) {
-      throw new UnknownHostException(
-          "BootstrapDns called for " + hostname + " instead of " + dnsHostname);
+  @Throws(UnknownHostException::class)
+  override fun lookup(hostname: String): List<InetAddress> {
+    if (this.dnsHostname != hostname) {
+      throw UnknownHostException(
+        "BootstrapDns called for $hostname instead of $dnsHostname"
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt]: Our Kotlin style is here too https://github.com/square/java-code-styles

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+ * <blockquote>A DNS API client encodes a single DNS query into an HTTP request
+ * using either the HTTP GET or POST method and the other requirements
+ * of this section.  The DNS API server defines the URI used by the
+ * request through the use of a URI Template.</blockquote>
+ *
+ * <h3>Warning: This is a non-final API.</h3>
+ *
+ *
+ * **As of OkHttp 3.11, this feature is an unstable preview: the API is subject to change,
+ * and the implementation is incomplete. We expect that OkHttp 3.12 or 3.13 will finalize this API.
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt]: nit: 4.0 or 4.1 ?

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+class DnsOverHttps internal constructor(builder: Builder) : Dns {
+  private val client: OkHttpClient
+  private val url: HttpUrl
+  private val includeIPv6: Boolean
+  private val post: Boolean
+  private val resolvePrivateAddresses: Boolean
+  private val resolvePublicAddresses: Boolean
+
+  init {
+    this.url = builder.url ?: throw NullPointerException("url not set")
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt]: nit: combine assignment with declaration?

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    this.includeIPv6 = builder.includeIPv6
+    this.post = builder.post
+    this.resolvePrivateAddresses = builder.resolvePrivateAddresses
+    this.resolvePublicAddresses = builder.resolvePublicAddresses
+    this.client = builder.client?.newBuilder()?.dns(buildBootstrapClient(builder))?.build()
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt]: nice

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    response.use {
+      if (!response.isSuccessful) {
+        throw IOException("response: " + response.code() + " " + response.message())
+      }
+
+      val body = response.body()
+
+      if (body!!.contentLength() > MAX_RESPONSE_SIZE) {
+        throw IOException(
+          "response size exceeds limit ("
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt]: nit: template?

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsRecordCodec.kt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+object DnsRecordCodec {
+  private const val SERVFAIL = 2
+  private const val NXDOMAIN = 3
+  const val TYPE_A = 0x0001
+  const val TYPE_AAAA = 0x001c
+  private const val TYPE_PTR = 0x000c
+  private val ASCII = StandardCharsets.US_ASCII
+
+  @JvmStatic
+  fun encodeQuery(host: String, type: Int): ByteString = Buffer().apply {
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsRecordCodec.kt]: ooh, nice

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsRecordCodec.kt [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      } else {
+        buf.skip(length.toLong())
+      }
+    }
+
+    return result
+  }
+
+  @Throws(EOFException::class)
+  private fun skipName(`in`: Buffer) {
```
Comment (swankjesse) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsRecordCodec.kt]: rename Buffer so source?

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt (hunk 2 above)
Comment (yschimke) [reply okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt]: Applied, was this comment about indent depth?

Comment (swankjesse) [reply okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/BootstrapDns.kt]: Yup!

**Pull request merged**